bakery/macaroon.go

```diff
@@ -0,0 +1,352 @@
+		for id64, data64 := range m1.CaveatData {
+			id, err := base64Decode([]byte(id64))
+			if err != nil {
+				return errgo.Notef(err, "cannot decode caveat id")
+			}
+			data, err := base64Decode([]byte(data64))
+			if err != nil {
+				return errgo.Notef(err, "cannot decode caveat")
+			}
+			caveatData[string(id)] = data
+		}
+		m.caveatData = caveatData
+		m.m = m1.Macaroon
+		m.namespace = m1.Namespace
+		// TODO should we allow version > LatestVersion here?
+		m.version = m1.Version
+		return nil
+	}
+	// Try to unmarshal from the original format.
+	var m2 *macaroon.Macaroon
+	if err := json.Unmarshal(data, &m2); err != nil {
+		return errgo.Mask(err)
+	}
+	m.m = m2
+	v, err := bakeryVersion(m2.Version())
+	if err != nil {
+		return errgo.Mask(err)
+	}
+	m.version = v
+	m.namespace = legacyNamespace()
+	return nil
+}
+
+// base64Decode base64-decodes the given data.
+// It accepts both standard padded encoding and unpadded
+// URL encoding.
+func base64Decode(data []byte) ([]byte, error) {
+	// Make a buffer that's big enough for both padded and
+	// unpadded cases.
+	buf := make([]byte, base64.RawStdEncoding.DecodedLen(len(data)))
+	if n, err := base64.StdEncoding.Decode(buf, data); err == nil {
+		return buf[0:n], nil
+	}
+	n, err := base64.RawURLEncoding.Decode(buf, data)
+	if err == nil {
+		return buf[0:n], nil
+	}
+	return nil, errgo.Mask(err)
+}
+
+// bakeryVersion returns a bakery version that corresponds to
+// the macaroon version v. It is necessarily approximate because
+// several bakery versions can correspond to a single macaroon
+// version, so it's only of use when decoding legacy formats
+// (in Macaroon.UnmarshalJSON).
+//
+// It will return an error if it doesn't recognise the version.
+func bakeryVersion(v macaroon.Version) (Version, error) {
+	switch v {
+	case macaroon.V1:
+		// Use version 1 because we don't know of any existing
+		// version 0 clients.
+		return Version1, nil
+	case macaroon.V2:
+		// Note that this could also correspond to Version3, but
+		// this logic is explicitly for legacy versions.
+		return Version2, nil
+	default:
+		return 0, errgo.Newf("unknown macaroon version when legacy-unmarshaling bakery macaroon; got %d", v)
+	}
+}
+
+// NewMacaroon creates and returns a new macaroon with the given root
+// key, id and location. If the version is more than the latest known
+// version, the latest known version will be used. The namespace is that
+// of the service creating it.
+func NewMacaroon(rootKey, id []byte, location string, version Version, ns *checkers.Namespace) (*Macaroon, error) {
+	if version > LatestVersion {
+		version = LatestVersion
+	}
+	m, err := macaroon.New(rootKey, id, location, MacaroonVersion(version))
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot create macaroon")
+	}
+	return &Macaroon{
+		m:         m,
+		version:   version,
+		namespace: ns,
+	}, nil
+}
+
+// M returns the underlying macaroon held within m.
+func (m *Macaroon) M() *macaroon.Macaroon {
```

Comment:
I'm not sure either. I also debated embedding *macaroon.Macaroon, but then it would be a mutable field. I also debated making the whole thing a transparent type with no private members.

I'm open to arguments any which way.